In [1]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

In [2]:
data = pd.read_csv('CVD_cleaned.CSV')
data

General_Health                  Checkup Exercise Heart_Disease  \
0                Poor  Within the past 2 years       No            No   
1           Very Good     Within the past year       No           Yes   
2           Very Good     Within the past year      Yes            No   
3                Poor     Within the past year      Yes           Yes   
4                Good     Within the past year       No            No   
...               ...                      ...      ...           ...   
308849      Very Good     Within the past year      Yes            No   
308850           Fair  Within the past 5 years      Yes            No   
308851      Very Good      5 or more years ago      Yes            No   
308852      Very Good     Within the past year      Yes            No   
308853      Excellent     Within the past year      Yes            No   

       Skin_Cancer Other_Cancer Depression  \
0               No           No         No   
1               No           No         No   
2               No           No         No   
3               No           No         No   
4               No           No         No   
...            ...          ...        ...   
308849          No           No         No   
308850          No           No         No   
308851          No           No        Yes   
308852          No           No         No   
308853          No           No         No   

                                          Diabetes Arthritis     Sex  \
0                                               No       Yes  Female   
1                                              Yes        No  Female   
2                                              Yes        No  Female   
3                                              Yes        No    Male   
4                                               No        No    Male   
...                                            ...       ...     ...   
308849                                          No        No    Male   
308850                                         Yes        No    Male   
308851  Yes, but female told only during pregnancy        No  Female   
308852                                          No        No    Male   
308853                                          No        No  Female   

       Age_Category  Height_(cm)  Weight_(kg)    BMI Smoking_History  \
0             70-74        150.0        32.66  14.54             Yes   
1             70-74        165.0        77.11  28.29              No   
2             60-64        163.0        88.45  33.47              No   
3             75-79        180.0        93.44  28.73              No   
4               80+        191.0        88.45  24.37             Yes   
...             ...          ...          ...    ...             ...   
308849        25-29        168.0        81.65  29.05              No   
308850        65-69        180.0        69.85  21.48              No   
308851        30-34        157.0        61.23  24.69             Yes   
308852        65-69        183.0        79.38  23.73              No   
308853        45-49        160.0        81.19  31.71              No   

        Alcohol_Consumption  Fruit_Consumption  Green_Vegetables_Consumption  \
0                       0.0               30.0                          16.0   
1                       0.0               30.0                           0.0   
2                       4.0               12.0                           3.0   
3                       0.0               30.0                          30.0   
4                       0.0                8.0                           4.0   
...                     ...                ...                           ...   
308849                  4.0               30.0                           8.0   
308850                  8.0               15.0                          60.0   
308851                  4.0               40.0                           8.0   
308852                  3.0               30.0 

In [3]:
df = data[~data.duplicated(keep = False)]

In [4]:
df.isnull().sum()

General_Health                  0
Checkup                         0
Exercise                        0
Heart_Disease                   0
Skin_Cancer                     0
Other_Cancer                    0
Depression                      0
Diabetes                        0
Arthritis                       0
Sex                             0
Age_Category                    0
Height_(cm)                     0
Weight_(kg)                     0
BMI                             0
Smoking_History                 0
Alcohol_Consumption             0
Fruit_Consumption               0
Green_Vegetables_Consumption    0
FriedPotato_Consumption         0
dtype: int64

## data preparation 

In [4]:
mapp = {
    "Yes": 1,
    "No" : 0
}


In [5]:
df["Heart_Disease"] = df["Heart_Disease"].map(mapp)

In [6]:
df = df.select_dtypes(exclude = 'object')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 308695 entries, 0 to 308853
Data columns (total 8 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   Heart_Disease                 308695 non-null  int64  
 1   Height_(cm)                   308695 non-null  float64
 2   Weight_(kg)                   308695 non-null  float64
 3   BMI                           308695 non-null  float64
 4   Alcohol_Consumption           308695 non-null  float64
 5   Fruit_Consumption             308695 non-null  float64
 6   Green_Vegetables_Consumption  308695 non-null  float64
 7   FriedPotato_Consumption       308695 non-null  float64
dtypes: float64(7), int64(1)
memory usage: 21.2 MB


In [7]:
# pisahkan target dan feature 
y = df['Heart_Disease']
X = df.drop("Heart_Disease", axis = 1)


In [22]:
X.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 308695 entries, 0 to 308853
Data columns (total 7 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   Height_(cm)                   308695 non-null  float64
 1   Weight_(kg)                   308695 non-null  float64
 2   BMI                           308695 non-null  float64
 3   Alcohol_Consumption           308695 non-null  float64
 4   Fruit_Consumption             308695 non-null  float64
 5   Green_Vegetables_Consumption  308695 non-null  float64
 6   FriedPotato_Consumption       308695 non-null  float64
dtypes: float64(7)
memory usage: 18.8 MB


In [23]:
y.info()

<class 'pandas.core.series.Series'>
Int64Index: 308695 entries, 0 to 308853
Series name: Heart_Disease
Non-Null Count   Dtype
--------------   -----
308695 non-null  int64
dtypes: int64(1)
memory usage: 4.7 MB


### under sampling

### split train, test, val

In [9]:
from sklearn.model_selection import train_test_split 
X_train, X_test,y_train,y_test = train_test_split(X,
                                                y,
                                                test_size = 0.2,
                                                random_state = 42)

In [15]:
y_train.value_counts()

0    255290
1     22535
Name: Heart_Disease, dtype: int64

In [11]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 277825 entries, 301955 to 122020
Data columns (total 7 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   Height_(cm)                   277825 non-null  float64
 1   Weight_(kg)                   277825 non-null  float64
 2   BMI                           277825 non-null  float64
 3   Alcohol_Consumption           277825 non-null  float64
 4   Fruit_Consumption             277825 non-null  float64
 5   Green_Vegetables_Consumption  277825 non-null  float64
 6   FriedPotato_Consumption       277825 non-null  float64
dtypes: float64(7)
memory usage: 17.0 MB


In [12]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30870 entries, 237911 to 9059
Data columns (total 7 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Height_(cm)                   30870 non-null  float64
 1   Weight_(kg)                   30870 non-null  float64
 2   BMI                           30870 non-null  float64
 3   Alcohol_Consumption           30870 non-null  float64
 4   Fruit_Consumption             30870 non-null  float64
 5   Green_Vegetables_Consumption  30870 non-null  float64
 6   FriedPotato_Consumption       30870 non-null  float64
dtypes: float64(7)
memory usage: 1.9 MB


In [14]:
y_test.value_counts()

0    28434
1     2436
Name: Heart_Disease, dtype: int64

In [10]:
from imblearn import under_sampling
X_under, y_under = under_sampling.RandomUnderSampler(sampling_strategy = 0.8).fit_resample(X_train, y_train)

from imblearn import over_sampling

# Overampling with smote
X_over_smote, y_over_smote = over_sampling.SMOTE().fit_resample(X_train, y_train)

In [17]:
y_over_smote.value_counts()

0    255290
1    255290
Name: Heart_Disease, dtype: int64

In [18]:
X_over_smote.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 510580 entries, 0 to 510579
Data columns (total 7 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   Height_(cm)                   510580 non-null  float64
 1   Weight_(kg)                   510580 non-null  float64
 2   BMI                           510580 non-null  float64
 3   Alcohol_Consumption           510580 non-null  float64
 4   Fruit_Consumption             510580 non-null  float64
 5   Green_Vegetables_Consumption  510580 non-null  float64
 6   FriedPotato_Consumption       510580 non-null  float64
dtypes: float64(7)
memory usage: 27.3 MB


In [42]:
X_under.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54727 entries, 0 to 54726
Data columns (total 7 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Height_(cm)                   54727 non-null  float64
 1   Weight_(kg)                   54727 non-null  float64
 2   BMI                           54727 non-null  float64
 3   Alcohol_Consumption           54727 non-null  float64
 4   Fruit_Consumption             54727 non-null  float64
 5   Green_Vegetables_Consumption  54727 non-null  float64
 6   FriedPotato_Consumption       54727 non-null  float64
dtypes: float64(7)
memory usage: 2.9 MB


In [41]:
y_under.value_counts()

0    32192
1    22535
Name: Heart_Disease, dtype: int64

# Modeling with KNN

In [11]:
''' Modeling menggunakan KNN'''

from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier()
knn.fit(X_under, y_under)

KNeighborsClassifier()

In [12]:
from sklearn.metrics import roc_auc_score, classification_report, f1_score
y_pred_train = knn.predict(X_under)
print(classification_report(y_under, y_pred_train, digits=6))

              precision    recall  f1-score   support

           0   0.724463  0.766635  0.744952     24978
           1   0.685412  0.635540  0.659535     19983

    accuracy                       0.708369     44961
   macro avg   0.704937  0.701087  0.702244     44961
weighted avg   0.707107  0.708369  0.706988     44961



In [13]:
y_pred_test = knn.predict(X_test)
print(classification_report(y_test, y_pred_test, digits=6))

              precision    recall  f1-score   support

           0   0.927605  0.623143  0.745486     56751
           1   0.094347  0.446672  0.155788      4988

    accuracy                       0.608886     61739
   macro avg   0.510976  0.534908  0.450637     61739
weighted avg   0.860284  0.608886  0.697843     61739



# Modeling with Logistic Regression

In [14]:
'''Modeling menggunakan Logreg'''

from sklearn.linear_model import LogisticRegression

logreg = LogisticRegression(random_state=42)
logreg.fit(X_under, y_under)
y_pred_train_log = logreg.predict(X_under)

In [15]:
print(classification_report(y_under, y_pred_train_log, digits=6))

              precision    recall  f1-score   support

           0   0.571485  0.876171  0.691764     24978
           1   0.536004  0.178802  0.268153     19983

    accuracy                       0.566224     44961
   macro avg   0.553744  0.527487  0.479958     44961
weighted avg   0.555715  0.566224  0.503489     44961



In [16]:
y_pred_test_log = logreg.predict(X_test)
print(classification_report(y_test,y_pred_test_log))

              precision    recall  f1-score   support

           0       0.92      0.88      0.90     56751
           1       0.11      0.17      0.13      4988

    accuracy                           0.82     61739
   macro avg       0.52      0.52      0.52     61739
weighted avg       0.86      0.82      0.84     61739



# Modeling with Decision Tree

In [17]:
'''modeling menggunakan decision tree'''

from sklearn.tree import DecisionTreeClassifier

dt = DecisionTreeClassifier(random_state=42)
dt.fit(X_under, y_under)
y_pred_train_dt = dt.predict(X_under)

In [ ]:
from sklearn.model_selection import RandomizedSearchCV

parameters = {
    'n_estimators': (10,20,30,40,50),
    'max_depth':(1,2,3,4,5)
}

rf_clf_randomcv = RandomizedSearchCV(rf_clf, parameters, cv=5, 
                                     scoring='recall', n_iter=10)
rf_clf_randomcv.fit(X_under, y_under)

In [18]:
print(classification_report(y_under,y_pred_train_dt))

              precision    recall  f1-score   support

           0       0.98      1.00      0.99     24978
           1       1.00      0.98      0.99     19983

    accuracy                           0.99     44961
   macro avg       0.99      0.99      0.99     44961
weighted avg       0.99      0.99      0.99     44961



In [19]:
y_pred_test_dt = dt.predict(X_test)
print(classification_report(y_test,y_pred_test_dt))

              precision    recall  f1-score   support

           0       0.93      0.58      0.71     56751
           1       0.09      0.47      0.15      4988

    accuracy                           0.57     61739
   macro avg       0.51      0.52      0.43     61739
weighted avg       0.86      0.57      0.67     61739



# RandomForest

In [22]:
'''modeling menggunakan random forest '''

from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier()
rf.fit(X_under,y_under)

RandomForestClassifier()

In [47]:
from sklearn.model_selection import RandomizedSearchCV

parameters = {
    'n_estimators': (10,20,30,40,50),
    'max_depth':(1,2,3,4,5)
}

rf_clf_randomcv = RandomizedSearchCV(rf, parameters, cv=5, 
                                     scoring='recall', n_iter=10)
rf_clf_randomcv.fit(X_under, y_under)

RandomizedSearchCV(cv=5, estimator=RandomForestClassifier(),
                   param_distributions={'max_depth': (1, 2, 3, 4, 5),
                                        'n_estimators': (10, 20, 30, 40, 50)},
                   scoring='recall')

In [51]:
cv_result = pd.DataFrame(rf_clf_randomcv.cv_results_)
retain_cols = ['params','mean_test_score','rank_test_score']
cv_result[retain_cols].sort_values('rank_test_score')

params  mean_test_score  rank_test_score
0  {'n_estimators': 20, 'max_depth': 5}           0.0004                1
5  {'n_estimators': 10, 'max_depth': 5}           0.0004                2
1  {'n_estimators': 40, 'max_depth': 3}           0.0000                3
2  {'n_estimators': 50, 'max_depth': 4}           0.0000                3
3  {'n_estimators': 20, 'max_depth': 2}           0.0000                3
4  {'n_estimators': 50, 'max_depth': 2}           0.0000                3
6  {'n_estimators': 50, 'max_depth': 1}           0.0000                3
7  {'n_estimators': 20, 'max_depth': 3}           0.0000                3
8  {'n_estimators': 10, 'max_depth': 2}           0.0000                3
9  {'n_estimators': 30, 'max_depth': 3}           0.0000                3

In [23]:
y_pred_train = rf.predict(X_under)
print(classification_report(y_under,y_pred_train))

              precision    recall  f1-score   support

           0       0.99      0.99      0.99     24978
           1       0.99      0.99      0.99     19983

    accuracy                           0.99     44961
   macro avg       0.99      0.99      0.99     44961
weighted avg       0.99      0.99      0.99     44961



In [26]:
y_pred_test = rf.predict(X_test)
print(classification_report(y_test,y_pred_test))

              precision    recall  f1-score   support

           0       0.93      0.68      0.78     56751
           1       0.10      0.43      0.17      4988

    accuracy                           0.66     61739
   macro avg       0.52      0.55      0.48     61739
weighted avg       0.86      0.66      0.73     61739



In [20]:
'''modeling menggunakan Xgboost '''
from sklearn.ensemble import GradientBoostingClassifier

xgd = GradientBoostingClassifier()
xgd.fit(X_under, y_under)
y_pred_xgd = xgd.predict(X_under)
print(classification_report(y_under,y_pred_xgd))

              precision    recall  f1-score   support

           0       0.63      0.74      0.68     24978
           1       0.58      0.45      0.50     19983

    accuracy                           0.61     44961
   macro avg       0.60      0.59      0.59     44961
weighted avg       0.60      0.61      0.60     44961



In [21]:
y_val_xgd = xgd.predict(X_test)
print(classification_report(y_test,y_val_xgd))

              precision    recall  f1-score   support

           0       0.94      0.74      0.82     56751
           1       0.12      0.42      0.19      4988

    accuracy                           0.71     61739
   macro avg       0.53      0.58      0.51     61739
weighted avg       0.87      0.71      0.77     61739



hasil baseline :
1. data yang ditrain dan test adalah data yang sudah di undersampling
2. data yang menjadi validasi akhir adalah data yang sudang dipisahkan untuk validaiton
3. karena menggunakan metrik recall, dari hasil pengujian nilai recall tertinggi ada pada model random forest yaitu sebesar 0.78 untuk memprediksi target sakit cardiovascular. patokan best model dari hasil prediksi x_val 